src/deluge/io/midi/midi_device_manager.cpp

```diff
@@ -530,17 +531,60 @@ void readAHostedDeviceFromFile() {
 
 void ConnectedUSBMIDIDevice::bufferMessage(uint32_t fullMessage) {
 	// If buffer already full, flush it
-	if (numMessagesQueued >= MIDI_SEND_BUFFER_LEN) {
-		midiEngine
-		    .flushUSBMIDIOutput(); // TODO: this is actually far from perfect - what if already sending - and if we want to wait/check for that, we should be calling the routine.
-		                           // And ideally, we'd be able to flush for just one device.
-		numMessagesQueued = 0;
+	// if (numMessagesQueued >= MIDI_SEND_BUFFER_LEN) {
+	// 	midiEngine
+	// 	    .flushUSBMIDIOutput(); // TODO: this is actually far from perfect - what if already sending - and if we want to wait/check for that, we should be calling the routine.
+	// 	                           // And ideally, we'd be able to flush for just one device.
+	// 	numMessagesQueued = 0;
+	// }
+	//
+	int queued = ringBufWriteIdx - ringBufReadIdx;
+	// if (queued > some_smaller_number) {
+	//   if (notAlreadySending) {
+	//     startTheSend();
+	//   }
+	// }
+	if (queued > MIDI_SEND_BUFFER_LEN_RING) {
+		// TODO: le panik
+		return;
 	}
 
-	preSendData[numMessagesQueued++] = fullMessage;
+	sendDataRingBuf[ringBufWriteIdx & MIDI_SEND_RING_MASK] = fullMessage;
+	ringBufWriteIdx++;
+
 	anythingInUSBOutputBuffer = true;
 }
 
+bool ConnectedUSBMIDIDevice::hasRingBuffered() {
+	// must me the same unsigned type as ringBufWriteIdx/ringBufReadIdx
+	uint32_t queued = ringBufWriteIdx - ringBufReadIdx;
+	return queued > 0;
+}
+
+bool ConnectedUSBMIDIDevice::consumeBytes() {
```

Comment:
Can you add an explanatory comment on what this functions purpose is and when it should be called?